src/object.c

```diff
@@ -17,18 +20,73 @@
 #define strtold(a,b) ((long double)strtod((a),(b)))
 #endif
 
+/* For objects with large embedded keys, we reserve space for an expire field,
+ * so if expire is set later, we don't need to reallocate the object. */
+#define KEY_SIZE_TO_INCLUDE_EXPIRE_THRESHOLD 128
+
 /* ===================== Creation and parsing of objects ==================== */
 
-robj *createObject(int type, void *ptr) {
-    robj *o = zmalloc(sizeof(*o));
+/* Creates an object, optionally with embedded key and expire fields. The key
+ * and expire fields can be omitted by passing NULL and -1, respectively. */
+kvobj *kvobjCreate(int type, const sds key, void *valptr, long long expire) {
+    /* Determine embedded key and expiration flags */
+    int iskvobj = key != NULL;
+    int has_expire = (expire != -1 ||
+                      (iskvobj && sdslen(key) >= KEY_SIZE_TO_INCLUDE_EXPIRE_THRESHOLD));
+    
+    /* Calculate embedded key size */
+    size_t key_sds_len = iskvobj ? sdslen(key) : 0;
+    char key_sds_type = iskvobj ? sdsReqType(key_sds_len) : 0;
+    size_t key_sds_size = iskvobj ? sdsReqSize(key_sds_len, key_sds_type) : 0;
+
+    /* Compute the base object size. Example of kvobj "mykey" and expiry:
+     *    +------+--------+--------------+--------------------+
+     *    | robj | expiry | key-hdr-size | sdshdr5 "mykey" \0 | 16+8+1+7 = 32bytes
+     *    +------+--------+--------------+--------------------+
+     */
+    size_t min_size = sizeof(robj);
+    if (has_expire) min_size += sizeof(long long);
+    if (iskvobj) min_size += 1 + key_sds_size; /* 1 byte for SDS header size */
+
+    /* Allocate object memory */
+    size_t bufsize = 0;
+    robj *o = zmalloc_usable(min_size, &bufsize);
     o->type = type;
     o->encoding = OBJ_ENCODING_RAW;
-    o->ptr = ptr;
+    o->ptr = valptr;
     o->refcount = 1;
     o->lru = 0;
+    o->iskvobj = iskvobj;
+
+    /* If extra space allows, pre-allocate anyway expiration */
+    if (iskvobj && !has_expire && bufsize >= min_size + sizeof(long long)) {
+        has_expire = 1;
+        min_size += sizeof(long long);
+    }
+    o->expirable = has_expire;
+
+    /* The memory after the struct where we embedded data. */
+    char *data = (void *)(o + 1);
+
+    /* Set the expire field. */
+    if (o->expirable) {
+        *(long long *)data = expire;
+        data += sizeof(long long);
+    }
+
+    /* Store embedded key. */
+    if (o->iskvobj) {
+        *data++ = sdsHdrSize(key_sds_type);
+        sdsnewplacement(data, key_sds_size, key_sds_type, key, key_sds_len);
+    }
+
     return o;
 }
 
+robj *createObject(int type, void *valptr) {
+    return kvobjCreate(type, NULL, valptr, -1);
+}
```

Comment:
what about keeping original createObject().
most of the paths of `kvobjCreate()` are actually useless for `createObject()`.
and there won't be too much duplica code for both of them, it's worth to make createObject() simple if the original one is faster.

Reply:
Nice catch! i will modify it.